### PR TITLE
rich probe failed message: add exitCode & err

### PR DESCRIPTION
**Pull request by NoicFank**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
add exitCode & err message, while probe failed, which could help us to troubleshot.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

**Comment by pacoxu** [comment]
Similar to #103602, the detailed error message may cause potential information exposure. https://github.com/kubernetes/kubernetes/pull/103602#discussion_r681254886 

/cc @endocrimes

**Comment by endocrimes** [comment]
/ok-to-test
/hold

I _think_ this should be safe enough on first pass - in that we already include stdout in the response. At the very least I'd be a proponent of including status code, even if we pass on the expansion of stderr.

I'm gonna reach out to our friends in sig-security for a second opinion.

**Comment by NoicFank** [comment]
> /ok-to-test /hold
> 
> I _think_ this should be safe enough on first pass - in that we already include stdout in the response. At the very least I'd be a proponent of including status code, even if we pass on the expansion of stderr.
> 
> I'm gonna reach out to our friends in sig-security for a second opinion.
@endocrimes @pacoxu 
The original intention of submitting this MR is that the probe failed in our production environment, which caused a series of problems.

At this time, we need to locate the reason why the exec probe failed. Then we open the kubelet debug log level, and we can  see the following information:
```
2022-06-21T01:10:40.791+0800 DEBUG xxx xxx/xxx.go: xxx Probe failed {"pod": "xxx", "podUID":
"xxx", "containerName": "xxxx", "probeType": "Readiness", "output": ""}}
```
Here, the probe output in the log is empty, which makes me need to exec into the corresponding pod and manually execute the script to check the reason.
```sh
[root@xxx /data/xxx]# ./probe-readiness.sh
[root@xxx /data/xxx]# echo $?
1
[root@xxx /data/xxx]#
```

So I expect to get the following tips directly through the log, especially exitCode:
```
2022-06-27T16:56:26.941+0800 DEBUG xxx xxx/xxx.go: xxx Probe failed {"pod": "xxx", "podUID":
"xxx", "containerName": "xxxx", "probeType": "Readiness", "output": "exitCode: 1, stderr: command terminated with exit code 1, stdout: "}}
```
ps: Due to company policies, some private information has been hidden

**Comment by NoicFank** [comment]
/ping @endocrimes please let me know, is there any suggestions?

**Review by endocrimes — COMMENTED** [review]
/hold-cancel
/lgtm

**Comment by pacoxu** [comment]
> /hold-cancel

/unhold
/triage accepted

**Comment by NoicFank** [comment]
/ping @tallclair @sjenning @endocrimes @pacoxu 

PTAL, thanks. 😊

File: pkg/probe/exec/exec.go
Comment (tallclair) [inline pkg/probe/exec/exec.go:70]: `data` here includes both stdout and stderr (lines 55 & 56 above). From the implementation, it looks like `exit.Error` will typically be a summary of why the process halted, and the exit status, making it redundant with the exit code. A better message might be to just append `exit.Error()` to the `string(data)`. E.g:

```suggestion
			probeFailedMsg := string(data)
			if probeFailedMsg[len(probeFailedMsg)-1] != '\n' {
			  probeFailedMsg += '\n'
			}
			probeFailedMsg += exit.Error()
```

File: pkg/probe/exec/exec_test.go
Comment (tallclair) [inline pkg/probe/exec/exec_test.go:134]: It would be nice to write these tests with real commands, to see what the actual output would look like. That probably requires isolating the tests per OS with build tags though.

File: pkg/probe/exec/exec.go
Comment (NoicFank) [reply pkg/probe/exec/exec.go]: @tallclair appreciate for your reviews. your suggestion is excellent, I have updated the code. 

Here is a small change, that exitCode could not appear in exit.Error(), thus I updated method exit.String() to return both exitCode & exitErrorMsg.

please take a look, thanks a lot.

File: pkg/probe/exec/exec_test.go
Comment (NoicFank) [reply pkg/probe/exec/exec_test.go]: yeah, I'm willing to do this, but I found the process exist error msg is  closely related to the error output of the user script, thus I pay more attention to the correctness of the program results in the test and updated the Exec test method to make it clearer and more comprehensive.

**Review by pacoxu — APPROVED** [review]
/lgtm

File: pkg/probe/exec/exec.go
Comment (bart0sh) [inline pkg/probe/exec/exec.go:71]: @NoicFank this is a bit unclear to me. We're trimming message in the check, but then we add "\n" to the original(not trimmed) message. So, for the string "Error: some error \n   " we'll add extra "\n" to the end.
Can you explain why is this needed?

Comment (bart0sh) [reply pkg/probe/exec/exec.go:71]: nvm, I understand it now. Makes sense to me.

Comment (NoicFank) [reply pkg/probe/exec/exec.go:71]: > nvm, I understand it now. Makes sense to me.

thanks for review 😊

**Comment by NoicFank** [comment]
ping @endocrimes @sjenning @pacoxu @tallclair 

PTAL, is there anything I could do to push it forward?

File: staging/src/k8s.io/client-go/util/exec/exec.go
Comment (tallclair) [inline staging/src/k8s.io/client-go/util/exec/exec.go:35]: (optional) if you just inline this, you don't need to wait for `client-go/util` approval.

File: pkg/probe/exec/exec_test.go
Comment (lavalamp) [inline pkg/probe/exec/exec_test.go:116]: I don't know anything about anything here but it seems super super confusing to make exiting 0 success have an *error* associated with it!!

File: staging/src/k8s.io/client-go/util/exec/exec.go
Comment (lavalamp) [reply staging/src/k8s.io/client-go/util/exec/exec.go:35]: I don't understand the purpose of this constructor? It is not doing anything that clients can't do for themselves?

**Pull request closed without merging**
